README.md
# STAT _(Unit-test Framework)_ <!-- omit in toc -->


![Programming C](https://img.shields.io/badge/programming-TDD%20|%20Embedded-orange.svg?style=flat&logo=c&logoColor=white) 
![MSVS](https://img.shields.io/badge/MSVS-2005--2008%20|%202010--2022-blue?style=flat&logo=Visual-Studio&logoColor=white) 
![GNU](https://img.shields.io/badge/GNU-Make%20|%20GCC-blue?style=flat&logo=GNU&logoColor=white) 
![Windows](https://img.shields.io/badge/Windows-7%20|%2010|%2011-blue?style=flat&logo=Windows&logoColor=white) 
![Linux](https://img.shields.io/badge/Linux-Debian%20Derivatives%20+-blue?style=flat&logo=Linux&logoColor=white) 
![python](https://img.shields.io/badge/python-3.5--3.11-blue?style=flat&logo=Python&logoColor=white)  
[![standard-readme compliant](https://img.shields.io/badge/readme_style-standard-brightgreen)](https://github.com/RichardLitt/standard-readme)
[![REUSE Compliance](https://github.com/westerndigitalcorporation/stat/workflows/REUSE%20Compliance/badge.svg?branch=master)](https://api.reuse.software/info/git.fsfe.org/reuse/api)
[![Regression](https://github.com/westerndigitalcorporation/stat/workflows/Regression/badge.svg?branch=master)](https://github.com/westerndigitalcorporation/stat/actions?query=workflow%3ARegression+branch%3Amaster)

<u>**St**</u>and<u>**A**</u>lone unit-<u>**T**</u>esting framework for software written in C

STAT is designed to promote an instant assimilation of Test-Driven Development in conditions of large-scale codebases that are constrained with *heavy load of legacy code*.
The framework was conceived with special regard to limitations of embedded environments and to implementation-concepts widely accepted in embedded programming.
It is based on a really great open-source project named [Unity](http://www.throwtheswitch.org/unity) (which can be found also on [GitHub](https://github.com/ThrowTheSwitch/Unity)).

## Table of Contents <!-- omit in toc -->

- [1. Background](#1-background)
- [2. Installation](#2-installation)
- [3. Usage](#3-usage)
  - [3.1. User-Guides](#31-user-guides)
  - [3.2. Additional Helpful Documentation](#32-additional-helpful-documentation)
- [4. Integration](#4-integration)
  - [4.1. OS](#41-os)
  - [4.2. Build Tools](#42-build-tools)
  - [4.3. IDEs](#43-ides)
- [5. Maintainer](#5-maintainer)
- [6. Contribution](#6-contribution)
  - [6.1. Contributors](#61-contributors)
  - [6.2. Imported Components](#62-imported-components)
  - [6.3. Unity Harness](#63-unity-harness)
  - [6.4. Contributing](#64-contributing)
- [9. License](#9-license)
- [8. Definitions](#8-definitions)

## 1. Background

We tried several unit-test frameworks, and there are quite a few really great products.
So, we stated the following principles as our requirements for the framework we were looking for:

- *Simplicity* – test-setup shall be simple, fast and intuitive
- *Speed* – execution shall be fast and focused to support TDD short cycles
- *Lightweight* – portability to embedded platforms shall be simple (planned for the future)
- *Comprehensive feedback* – better logging &rArr; lesser step-by-step debugging &rArr; better efficiency  
- *Reproducibility*– tests shall be reproducible and deterministic
- *Test-code sharing* – reduce the inevitable code-duplication (e.g. test-doubles)
- *Automation* – test automation shall be very simple to achieve
- <u>***CUT-Isolation***</u> – decouple from noise-impact of other FW-components, HW or OS

_The last one we found most critical for our needs due to constraints described in [Conceptual Model](./docs/conceptual_model.md)._ The *CUT-Isolation* principle was the one that we couldn't achieve with existing alternative frameworks.

In addition, we had certain limitations that hardened integration of other solutions in our environment. Our lab machines were limitted to run only *Windows* OS, *Python* 2.7, *MS Visual Studio* and our target build tool-chain. No lua, Ruby, CMake or any other things of that kind.

Eventually, it was decided to build our own framework that will meet all these requirements. But we didn't do everythign from scratch. We derived our solution from [Unity Harness](#63-unity-harness).

## 2. Installation

Please refer the [Getting Started](./docs/stat_getting_started.md) user-guide. It describes how to install and to setup the framework, and how to compose unit-tests over it.

## 3. Usage

### 3.1. User-Guides

- [Command-line](docs/stat_commandline.md) - full description of all possible command-line arguments to control behavior of STAT
- [Unity Assertions Reference](./unity/docs/UnityAssertionsReference.md) - the original '_Unity_' user-guide with usage-description of assertion macros
- [STAT-Mock Library](./docs/stat_mock.md) - user-guide describing how to use built-in Mock library of STAT framework

### 3.2. Additional Helpful Documentation

- [Conceptual Model](./docs/conceptual_model.md) - conceptual model underlying the design of the framework, a valuable knowledge to become Pro with STAT
- [Unity Assertions Cheat Sheet](./unity/docs/UnityAssertionsCheatSheetSuitableforPrintingandPossiblyFraming.pdf) - a cook-book with macros worth printing and framing
- [Unity Configuration Guide](./unity/docs/UnityConfigurationGuide.md) - the default configuration of Unity defined within STAT can be overwritten

## 4. Integration

### 4.1. OS

STAT-framework is built to run both on **Windows** and **Linux**.

### 4.2. Build Tools

STAT-framework has an integrated support for build-tools, which are determined based on OS, localy instlled tools and user-configuration:

- On Windows: _Microsoft Visual Studio_ including all versions from 2005 up to 2019 (any edition, i.e. Community, Professional and Enterprise).  

- On Linux: _GNU Compiler_, i.e. GCC

STAT is designed to minimize its requirements from the system. Therefore, it uses a precompiled _GNU Make Tool_ (included in the repo).

> This decision was dictated by restrictions of our lab, which didn't allow installing any additional software.

***MSVS Versioning***

If not explicitly specified in file "[.statconfig](./docs/statconfig.md)", the latest MSVS version on the machine gets identified and gets set as active build-tools. Otherwise, the specified version is selected instead. The latter is useful to fixate on common single version for all developer and lab machines.

### 4.3. IDEs

Framework provides services to generate project files for certain IDEs, by the means of [command-line](docs/stat_commandline.md). The following IDEs are currently supported:

- _Microsoft Visual Studio_ of version corresponding to the selected [build-tools](#42-build-tools)
  - The generated project file for this IDE supports step-by-step debugging
- _Source Insight 4.0_ - it is a unique IDE, which also is a code-analyzer, actually one of a kind
- _Visual-Studio Code_ - the generated workspace for this framework supports step-by-step debugging on Linux

## 5. Maintainer

&nbsp;&nbsp;&nbsp;*Arseniy Aharonov* - the maintainer and the main contributor  
&nbsp;&nbsp;&nbsp;[![GitHub](https://img.shields.io/badge/%20-@are&minus;scenic-blue.svg?style=social&logo=GitHub)](https://github.com/are-scenic) [![GitHub](https://img.shields.io/badge/arseniy@aharonov.icu-blue.svg?style=social&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAANCAYAAACgu+4kAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAFZSURBVDhPlZK7S8NgFMXz5dE8WwSRYJcMTrqICDo7BBdBiy4OdRb/AsFRR0Fwc9XNVXfdnESQgo+Cg4K6qIMUB2Pi74ag0KrEA4fv5txzz/cgyjTNvSRJJjRN+4Cloeu6YVnWvqK4hk1CHopeKTAznWXZpqaUerZte6rQS4OTr7K8S52S1gqCYFg+SkD5vh+z8T11Jid4gqeEHJM6iWjlth8QRZHDvRfx3+A/R8pP0EIcNQxjnUabkBW0npBqtdqPZ4vBM/xLjuM0kTvSa8FBSa9UKmsYrniTZbQ+aQKdI4cMbdNr45lHM1hnWb8DpACKxgw7XcBd3mWEsJjBI3joed6YeMT4W4BAr9VqQwSccKU7eMu1DggboJcPC/4KyOG6bp2QBozDMPQL+QvdAXUo6d3UC/b0CJhj7cifeJmm6Q4fL7A0uNo4bGg80gbFK9rbf8jGj67rLnwCZYNQShjAlXoAAAAASUVORK5CYII=)](mailto:arseniy@aharonov.icu)

## 6. Contribution

### 6.1. Contributors

An open source project thrives on contribution of different people, and it may come in many forms. Some develope the very code, others provide constructive feedback, and some share wisdom through guidance and consultation. And all of these are very valuable contributions made by very valuable contributors:  

- Eitan Talianker  
[![GitHub](https://img.shields.io/badge/@etalianker-blue.svg?style=social&logo=GitHub)](https://github.com/etalianker)
- Dr. Eitan Farchi  
[![GitHub](https://img.shields.io/badge/farchi@il.ibm.com-blue.svg?style=social&logo=IBM)](mailto:farchi@il.ibm.com)
- Udi Arie  
[![GitHub](https://img.shields.io/badge/@udiarie-blue.svg?style=social&logo=GitHub)](https://github.com/udiarie)
- Oran DeBotton  
[![GitHub](https://img.shields.io/badge/SolarEdge-oran.debotton@SolarEdge.com-blue.svg?style=social)](mailto:oran.debotton@SolarEdge.com)
- Luna Benarroch  
[![GitHub](https://img.shields.io/badge/@luna--benarroch-blue.svg?style=social&logo=GitHub)](https://github.com/luna-benarroch) [![GitHub](https://img.shields.io/badge/luna.benarroch@ibm.com-blue.svg?style=social&logo=IBM)](mailto:luna.benarroch@ibm.com)
- Leah Adler  
[![GitHub](https://img.shields.io/badge/@leahadler-blue.svg?style=social&logo=GitHub)](https://github.com/leahadler)

### 6.2. Imported Components

STAT wasn't build from scratch. Several valuable components underlay the framework:

- Unity harness source-files - a [key component](#63-unity-harness) for the framework
  - License: [MIT](./unity/LICENSE.txt)
  - Source: [on GitHub](https://github.com/ThrowTheSwitch/Unity)
- Visual Studio Locator (a.k.a. `vswhere`) - a utility to locate Visual Studio in newer versions of MSVS.
  - License: [MIT](https://github.com/microsoft/vswhere/blob/ded0fdd04473772af1dd001d82b6e3c871731788/LICENSE.txt)
  - Source: [on GitHub](https://github.com/Microsoft/vswhere/releases)
- GNU Make tool - controls generation of executables from sources based on makefiles
  - License: [GPLv3+ license](LICENSES/GPL-3.0-or-later.txt)
  - Source: [on GNU.org](https://www.gnu.org/software/make/)

### 6.3. Unity Harness

Though we couldn't find something that will answer all our requirements, we found the one that was very close.  
[*Unity harness*](http://www.throwtheswitch.org/unity) is that great "almost"-match with the following clear advantages:

- Minimalistic in size and dependencies
- Can be compiled almost on any platform
- Provides rich and strong assertion mechanism
- Prints very comprehensive log and results
- Tolerant to test-failures
  - Failing test (if properly built) doesn't crash the subsequent tests

An additional advantage worthy of a separate mentioning is the very fact that Unity is written in C, which is the same language we use for our production code.  
Writing in one language gives the same sense of developing, no matter whether it is the production code or unit-tests, without constant switching between different language paradigms.

### 6.4. Contributing

Feel free to impact! [Open an issue](https://github.com/westerndigitalcorporation/stat/issues).  

<!--
STAT framework follows the [Contributor Covenant](http://contributor-covenant.org/version/1/3/0/) Code of Conduct.
-->

## 9. License

The base-code of the STAT framework is licensed under [MIT license](LICENSES/MIT.txt). In addition, this repo contains a binary of GNU-Make tool, which is distributed under [GPLv3+ license](LICENSES/GPL-3.0-or-later.txt).

## 8. Definitions

_These definitions are provided to clarify terms used above._

- **STAT** – stands for standalone testing framework
- **CUT** – means Code Under Test; usually it is a module or a sub-module that is tested
- **DOC** – means Depended-on Component; in the literature this term is used to describe external dependencies that our CUT depends on
- **CUT-Isolation** – is a method of decoupling the CUT from the other parts of FW-Code for unit-testing purposes, meaning how we deal with DOCs in our unit-tests
- **FW** - Firmware
- **Test-Double** – is a substitution of an operational version of DOC with a test-version that emulates a real interface, but serves testing goals
- **Test-package** – in terms of STAT-framework, it represents a group of unit-tests that run as a single executable; it usually covers a specific CUT
- **TDD** - _[Test-Driven Development](https://www.agilealliance.org/glossary/tdd)_, see also _[TDD](https://martinfowler.com/bliki/TestDrivenDevelopment.html)_ for more details


CHANGELOG.md
# Changelog

[![The format is based on Keep a Changelog](https://img.shields.io/badge/Keep%20a%20Changelog-1.0.0-brightgreen?style=plastic)](https://keepachangelog.com/en/1.0.0/)
[![The project adheres to](https://img.shields.io/badge/Semantic%20Versioning-v2.0.0-brightgreen?style=plastic)](https://semver.org/spec/v2.0.0.html)
[![Date Format follows ISO-8601](https://img.shields.io/badge/ISO-8601%20Date%20Format%20-brightgreen?style=plastic)](http://www.iso.org/iso/home/standards/iso8601.htm)

All notable changes to this project will be documented in this file.

<!-- Types of changes:
### Added       - for new features.
### Changed     - for changes in existing functionality.
### Deprecated  - for soon-to-be removed features.
### Removed     - for now removed features.
### Fixed       - for any bug fixes.
### Security    - in case of vulnerabilities.

Release name format: [2.2.1] - 2020-02-02
-->

## [Unreleased]

None

## [2.2.0] - 2023-08-27

### Added

- Added support for MSVC 2022

### Fixed

- Fixed CI issues with non-compliant text in `dep5` file

### Removed

- Removed official support of Python 2.7 as its support was dropped by GitHub containers


## [2.1.7] - 2021-04-13

### Added

- Added support for double-versioning of dummy-interfaces
  - It allows more flexibility, though usually signals bad test-design

### Changed

- Improved performance of rebuild (-c) and 'gear'-ed (-g) command-line configurations.

### Fixed 

- Fixed an issue with the performance of the framework on Windows platform.

None

## [2.1.6] - 2021-04-03

### Fixed

- Fixed an issue due to which dummy interfaces always overwrote production headers even when it was not required

## [2.1.5] - 2021-03-31

### Changed

- Changed ability to run an ignored makefile if specified explicitly

### Fixed

- Fixed an issue with Spied-Data getter to assert whether the call was issued instead of returning null if it wasn't
- Fixed an issue with erroneous handling of build failures upon incremental build

## [2.1.4] - 2021-01-21

### Added

- Completed implementation of new Mock type Infinite-Mock

### Fixed

- Integrated Overridden Mock into functionality of Order-Tracking
- Fixed creation of Source-Insight project creation

## [2.1.3] - 2020-12-28

### Changed

- Improved incremental building - based on concepts by [Mad Scientist](http://make.mad-scientist.net/papers/advanced-auto-dependency-generation/)
- Optimized timing of rebuild
- Better encapsulation of Make definitions divided between OS and Tool-Chain

## [2.1.2] - 2020-12-12

### Fixed

- In REH-based Linux the VS-Code workspace had issues with relative paths used for CWD of build-tasks

## [2.1.1] - 2020-11-24

### Added 

- Added support for VS-Code on Windows Platform with full IDE integration

### Fixed

- Fixed the issue with precedence of header-files that are overwritten with test-double

## [2.1.0] - 2020-11-19

### Added

- Added support for GCC compiler on Linux Platform
- Added support for VS-Code on Linux Platform to enable debugging

### Changed

- Updated the documentation with the proper guidelines following the intriduced porting and support for the Linux environment

## [2.0.0-alpha] - 2020-10-15

### Added

- Added a precompiled version of GNU Make tool for Windows platform

### Changed

- Migrated build scripts from NMAKE to GNU Make tool

## [1.2.3] - 2020-08-29

### Added

- Added the initial list of contributors

### Changed

- Performed some refactoring in NMAKE script-structure to unify the method of compiling and linking both in full and incremental builds

### Fixed

- Fixed incorrect logging in case of internal failure of a test-package

## [1.2.2-alpha] - 2020-08-24

### Fixed 

- Fixed an issue related to inconsistency of NMAKE to deal with assignments of environment variables, which caused occasional failures of incremental build

## [1.2.1] - 2020-08-17

### Fixed 

- Fixed issues with incremental build that caused failures due to extreme over-checking of warnings fired by the linker

### Changed

- Improved the performance of the incremental build mode to ensure efficient running for automation
- Added a rule to MSVS core-makefiles that bypasses an issue with makefiles having bad syntax that caused incremental build to fail

## [1.2.0] - 2020-08-12

### Added

- Added several types of building options (e.g. incremental build, rebuild)

### Changed

- Modified the way that tool's root-path is referred to allow more flexibility
  - Supports execution of STAT-own unit-testfrom different IDEs (e.g. VS Code, PyCharm)
  - Reduces full-path references to minimum-> replaced (where possible) with relative path

### Fixed 

- Fixed the limitation within MSVS IDE that prevented compilation errors from being clickable from the debugger

## [1.1.0] - 2020-07-05

### Added

- Added support for Python 3 series (verified on versions 3.5 - 3.8, using Conda)
- Added support to several versions of Python Mock module in unit-tests of the framework

### Changed

- Changed processor counting method so that it doesn't mandatory requires `psutil` package
- Made additional adjustments to handling of `multiprocessing`-related flows, based on:
  - Issues discovered during Python 3 adjustments  
    and
  - Issues reported by other developers in the community
- Code-style changes to meet pep-8 (except for naming conventions)

## [1.0.0] - 2020-06-26

### Added

- First official release of STAT-framework to the open-source
- Added CHANGELOG.md following format defined by "[Keep a Changelog](https://keepachangelog.com/en/1.0.0/)"
- Added licensing information for the documentation of the framework
- Finished first draft of STAT-Framework documentation
- Added complete support to multiple versions of MS Visual Studio
- Added support for [Source Insight IDE](https://www.sourceinsight.com/)
- Added an option to boost performance by multiprocessing
- Added ability to force order-tracking for consumption of Mock-objects

services.py
#!/usr/bin/env python
# SPDX-FileCopyrightText: (c) 2020 Western Digital Corporation or its affiliates,
#                             Arseniy Aharonov <arseniy@aharonov.icu>
#
# SPDX-License-Identifier: MIT

from __future__ import print_function

import os
import platform
import subprocess
import sys
from fnmatch import filter as filterFileNames
from shutil import rmtree
from threading import Thread
from time import sleep
from json import dump as dumpJson
from shlex import split as splitCmdLine

import stat_attributes as attributes


def countCpuCores():
    try:
        from psutil import cpu_count
    except ImportError:
        try:
            from multiprocessing import cpu_count
        except ImportError:
            return 1
    try:
        return cpu_count()
    except NotImplementedError:
        return 1


def isWindows():
    return True if platform.system() == "Windows" else False


def getPlatform():
    # 'Linux32/64', 'Darwin32/64', 'Java32/64', 'Windows32/64'
    machine = "64" if platform.machine()[-2:] == "64" else "32"
    platformMode = platform.system() + machine
    return platformMode


def toWindowsPath(path):
    return path.replace('/', '\\')


def toPosixPath(path):
    return path.replace('\\', '/')


def toNativePath(path):
    return toWindowsPath(path) if isWindows() else toPosixPath(path)


def nameExecutable(fileOnlyName):
    return fileOnlyName + '.exe' if isWindows() else fileOnlyName


def mkdir(path, exist_ok=False):
    if exist_ok and os.path.isdir(path):
        return
    os.makedirs(path)


def formatCommandLine(command):
    if isWindows():
        return " ".join(command) if isinstance(command, (list, tuple)) else command
    else:
        return command if isinstance(command, (list, tuple)) else splitCmdLine(command)


def execute(command, beSilent=False, **kwargs):
    commandLine = formatCommandLine(command)
    arguments = dict(bufsize=1, universal_newlines=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    arguments.update(kwargs)
    process = subprocess.Popen(commandLine, **arguments)
    lines = []
    thread = Thread(target=__captureOutputLines, args=(process, beSilent, lines))
    thread.setDaemon(True)
    thread.start()
    thread.join()
    process.communicate()
    return process.returncode, lines


def __captureOutputLines(process, beSilent, lines):
    for _line in iter(process.stdout.readline, b''):
        if _line == '':
            break
        lines.append(_line)
        if not beSilent:
            print(_line, end='')


def executeForOutput(command, **kwargs):
    commandLine = formatCommandLine(command)
    arguments = dict(bufsize=1, universal_newlines=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    arguments.update(kwargs)
    process = subprocess.Popen(commandLine, **arguments)
    output, _ = process.communicate()
    return output.strip()


def remove(path):
    doesExist, attemptRemoval, typeName = \
        (os.path.isdir, rmtree, 'directory') if os.path.isdir(path) else (os.path.isfile, os.remove, 'file')
    while doesExist(path):
        try:
            attemptRemoval(path)
        except OSError:
            print("Waiting for {type} '{path}' to find unlocked (probably by AntiVirus)!".format(type=typeName,
                                                                                                 path=path))
            sleep(1)


def writeJsonFile(filePath, data):
    with open(filePath, 'w') as fp:
        dumpJson(data, fp, indent=3)


def createLink(sourcePath, targetPath):
    source = os.path.relpath(sourcePath, os.path.dirname(targetPath))
    target = targetPath

    if isWindows():
        source = toWindowsPath(source)
        target = toWindowsPath(target)
        if os.path.isdir(source):
            commandLine = 'cmd /c mklink /D "{target}" "{source}" >nul'
        else:
            commandLine = 'cmd /c mklink "{target}" "{source}" >nul'
        subprocess.Popen(formatCommandLine(commandLine.format(target=target, source=source)), shell=True).wait()
    else:
        os.symlink(source, target)  # pylint: disable=no-member


def findSubFolderOnPath(subFolder, path='.'):
    currentPath = os.getcwd() if path == '.' else path
    subFolderPath = os.path.join(currentPath, subFolder)
    while not os.path.isdir(subFolderPath):
        currentPath = os.path.dirname(currentPath)
        if currentPath == os.path.dirname(currentPath):
            return None
        subFolderPath = os.path.join(currentPath, subFolder)
    return subFolderPath


def getFileLocationThroughoutCurrentPath(fileName, currentPath='.'):
    previousDirectory = None
    currentDirectory = currentPath if currentPath != '.' else os.getcwd()
    while previousDirectory != currentDirectory:
        fileLocation = os.path.join(currentDirectory, fileName)
        if os.path.isfile(fileLocation):
            return fileLocation
        previousDirectory = currentDirectory
        currentDirectory = os.path.dirname(previousDirectory)
    return None


def listFilenames(pathName, encompassingPattern, *patternsForSubsets):
    files = filterFileNames(os.listdir(pathName), encompassingPattern)
    return __selectFilesByPatterns(files, patternsForSubsets) if patternsForSubsets else set(files)


def listMakefiles(pathName, *patterns):
    files = listFilenames(pathName, '*.mak', *patterns)
    if len(patterns) != len(files) or set(patterns) != files:
        ignored = __selectIgnoredFiles(files, pathName)
        if ignored:
            files.difference_update(ignored)
    return sorted(files)


def readFileLines(filepath):
    with open(filepath, 'r') as _file:
        return _file.readlines()


def readNonEmptyLines(filepath):
    for line in readFileLines(filepath):
        line = line.strip()
        if line:
            yield line


def readTextFileAtOnce(filepath):
    with open(filepath, 'r') as aFile:
        text = aFile.read()
    return text


def locateFile(filename, mode=None, path=None):
    def _locateFileOnPath(_filename, _path):
        _filepath = os.path.join(_path, _filename)
        if isWindows() and mode and mode & os.X_OK:
            _, _ext = os.path.splitext(_filepath)
            exeExtensions = os.environ.get("PATHEXT", "").split(os.pathsep)
            if not _ext:
                return _matchExtensionToFile(_filepath, exeExtensions)
            elif _ext not in exeExtensions:
                return ""
        return _filepath if _isFileAccessible(_filepath) else ""

    def _matchExtensionToFile(_filepath, _extensions):
        for _ext in _extensions:
            if _isFileAccessible(_filepath + _ext):
                return _filepath + _ext
        else:
            return ""

    def _isFileAccessible(_filepath):
        return os.path.isfile(_filepath) and (not mode or os.access(_filepath, mode))

    if isinstance(path, str):
        return _locateFileOnPath(filename, path)
    else:
        paths = [os.path.abspath(".")] + os.environ.get("PATH", "").split(os.pathsep) if not path else path
        for _path in paths:
            filepath = _locateFileOnPath(filename, _path)
            if filepath:
                return filepath
        else:
            return ""


def locateResource(filename):
    resource = os.path.join(attributes.TOOL_PATH, attributes.RESOURCES_DIRECTORY, filename)
    if not os.path.isfile(resource):
        raise ServicesException(ServicesException.RESOURCE_NOT_FOUND.format(resource))
    return resource


def formatMakeCommand(makefile, args=(), **variables):
    makeTool = attributes.MAKE_TOOL.get(getPlatform(), "make")
    command = [makeTool, "-f", makefile]
    command.extend(args)
    command.extend(['{0}="{1}"'.format(*pair) for pair in variables.items()])
    return command


def __selectIgnoredFiles(makefiles, pathName):
    ignoreFilePathname = os.path.join(pathName, attributes.IGNORE_FILENAME)
    if os.path.isfile(ignoreFilePathname):
        ignoreList = [line.strip() for line in readNonEmptyLines(ignoreFilePathname)]
        if ignoreList:
            return __selectFilesByPatterns(makefiles, ignoreList)
    return set()


def __selectFilesByPatterns(makefiles, patterns):
    filtered = []
    for pattern in patterns:
        filtered.extend(filterFileNames(makefiles, pattern))
    return set(filtered)


def meta_class(mcs, *cls, **mewAttributes):
    mangledName = "_".join(("_", mcs.__name__,) + tuple(item.__name__ for item in cls))
    return mcs(mangledName, cls, mewAttributes)


def abstract_method(method):
    return abstract_callable(method)


def abstract_callable(callableObject):
    def call_wrapper(*args, **kwargs):
        typeName = 'method' if args and hasattr(args[0], callableObject.__name__) else callableObject.__class__.__name__
        raise NotImplementedError("The {0} '{1}' must be implemented!".format(typeName, callableObject.__name__))
    return call_wrapper


class SingletonMeta(type):
    """ A metaclass that creates a Singleton base class when called. """
    __instances = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls.__instances:
            cls.__instances[cls] = super(SingletonMeta, cls).__call__(*args, **kwargs)
        return cls.__instances[cls]

    def clear(cls):
        try:
            del cls.__instances[cls]  # pylint: disable=no-member
        except KeyError:
            pass


class FactoryByLegacy(type):

    def __init__(cls, name, bases, spec):
        super(FactoryByLegacy, cls).__init__(name, bases, spec)
        if not hasattr(cls, '_announcedFactoryItems'):
            cls._announcedFactoryItems = {}
        cls.__registerIfAnnounced()

    def __registerIfAnnounced(cls):
        uidAttribute = getattr(cls, 'uidAttribute', 'UID')
        uidValue = getattr(cls, uidAttribute, None)
        if uidValue:
            cls._announcedFactoryItems[uidValue] = cls

    def __iter__(cls):
        for _uid in cls._announcedFactoryItems:
            yield _uid

    def __contains__(cls, uid):
        return uid in cls._announcedFactoryItems

    def __len__(cls):
        return len(cls._announcedFactoryItems)

    def get(cls, uid, default=None):
        return cls._announcedFactoryItems.get(uid, default)

    def create(cls, uid, *args, **kwargs):
        _cls = cls._announcedFactoryItems.get(uid, cls)
        return _cls(*args, **kwargs)


class Configuration(object):

    def __init__(self, **kwargs):
        super(Configuration, self).__init__()
        self.__configuration = kwargs

    def __iter__(self):
        for key in self.__configuration:
            yield key

    def __getitem__(self, key):
        return self.__configuration.get(key, None)

    def update(self, iterable=None, **kwargs):
        if iterable:
            self.__configuration.update({key: iterable[key] for key in iterable})
        self.__configuration.update(kwargs)

    def getInt(self, key, default=0):
        try:
            return int(self.__configuration.get(key, ''))
        except ValueError:
            return default

    def get(self, key, default):
        return self.__configuration.get(key, default)


class ServicesException(Exception):
    """
    Custom exception for STAT services
    """
    NO_VS_TOOLS_FOUND = "No VS TOOLS were found on this PC."
    NO_NMAKE_FOUND = "No NMAKE was found on this PC."
    RESOURCE_NOT_FOUND = "No resource by filename '{0}' was found."


if __name__ == '__main__':
    pass
